File: autocontext/src/autocontext/execution/rubric_spec.py
from __future__ import annotations

import re
from collections import Counter, defaultdict
from dataclasses import dataclass, field
from statistics import mean
from typing import Any, Literal, cast

RUBRIC_SPEC_SCHEMA_VERSION = 1
_FINDING_SEVERITY = Literal["warning", "error"]
_SCALE_KIND = Literal["numeric", "binary"]
_GENERIC_WORDS = {"good", "nice", "appropriate", "adequate", "proper"}
_HYPOTHESIS_LOADED = ("prove that", "confirm that", "must show", "obviously")


@dataclass(frozen=True, slots=True)
class RubricScope:
    include: list[str] = field(default_factory=list)
    exclude: list[str] = field(default_factory=list)

    @classmethod
    def model_validate(cls, data: dict[str, Any] | None) -> RubricScope | None:
        if not data:
            return None
        return cls(include=_string_list(data.get("include")), exclude=_string_list(data.get("exclude")))

    def has_boundary(self) -> bool:
        return bool(self.include or self.exclude)


@dataclass(frozen=True, slots=True)
class CorpusProfile:
    domain: str = ""
    audience: str = ""
    source_summary: str = ""

    @classmethod
    def model_validate(cls, data: dict[str, Any] | None) -> CorpusProfile | None:
        if not data:
            return None
        return cls(
            domain=str(data.get("domain", "")),
            audience=str(data.get("audience", "")),
            source_summary=str(data.get("source_summary", "")),
        )


@dataclass(frozen=True, slots=True)
class RubricScale:
    id: str
    kind: _SCALE_KIND = "numeric"
    min_score: float = 0.0
    max_score: float = 1.0
    pass_score: float | None = None
    anchors: dict[str, str] = field(default_factory=dict)

    @classmethod
    def model_validate(cls, data: dict[str, Any]) -> RubricScale:
        kind = data.get("kind")
        if kind not in ("numeric", "binary"):
            raise ValueError("scale kind must be 'numeric' or 'binary'")
        return cls(
            id=str(data["id"]),
            kind=cast(_SCALE_KIND, kind),
            min_score=float(data.get("min_score", 0.0)),
            max_score=float(data.get("max_score", 1.0)),
            pass_score=_optional_float(data.get("pass_score")),
            anchors={str(key): str(value) for key, value in dict(data.get("anchors", {})).items()},
        )


@dataclass(frozen=True, slots=True)
class RubricCriterion:
    id: str
    description: str
    scale_id: str
    weight: float = 1.0
    scope: RubricScope | None = None
    evidence_requirements: list[str] = field(default_factory=list)

    @classmethod
    def model_validate(cls, data: dict[str, Any]) -> RubricCriterion:
        if "scale_id" not in data:
            raise ValueError("criterion scale_id is required")
        return cls(
            id=str(data["id"]),
            description=str(data["description"]),
            scale_id=str(data["scale_id"]),
            weight=float(data.get("weight", 1.0)),
            scope=RubricScope.model_validate(_optional_dict(data.get("scope"))),
            evidence_requirements=_string_list(data.get("evidence_requirements")),
        )


@dataclass(frozen=True, slots=True)
class RubricDisqualifier:
    id: str
    description: str

    @classmethod
    def model_validate(cls, data: dict[str, Any]) -> RubricDisqualifier:
        return cls(id=str(data["id"]), description=str(data["description"]))


@dataclass(frozen=True, slots=True)
class DecisionThresholds:
    pass_score: float = 0.8
    excellent_score: float = 0.9

    @classmethod
    def model_validate(cls, data: dict[str, Any] | None) -> DecisionThresholds | None:
        if not data:
            return None
        return cls(pass_score=float(data.get("pass_score", 0.8)), excellent_score=float(data.get("excellent_score", 0.9)))


@dataclass(frozen=True, slots=True)
class RubricSpec:
    rubric_id: str
    goal: str
    criteria: list[RubricCriterion]
    scales: list[RubricScale]
    schema_version: int = RUBRIC_SPEC_SCHEMA_VERSION
    title: str = ""
    metadata: dict[str, Any] = field(default_factory=dict)
    scope: RubricScope | None = None
    corpus_profile: CorpusProfile | None = None
    disqualifiers: list[RubricDisqualifier] = field(default_factory=list)
    evidence_requirements: list[str] = field(default_factory=list)
    output_constraints: list[str] = field(default_factory=list)
    decision_thresholds: DecisionThresholds | None = None

    @classmethod
    def model_validate(cls, data: dict[str, Any]) -> RubricSpec:
        return cls(
            schema_version=int(data.get("schema_version", RUBRIC_SPEC_SCHEMA_VERSION)),
            rubric_id=str(data["rubric_id"]),
            title=str(data.get("title", "")),
            goal=str(data["goal"]),
            metadata=dict(data.get("metadata", {})),
            scope=RubricScope.model_validate(_optional_dict(data.get("scope"))),
            corpus_profile=CorpusProfile.model_validate(_optional_dict(data.get("corpus_profile"))),
            criteria=[RubricCriterion.model_validate(_require_dict(item)) for item in data.get("criteria", [])],
            scales=[RubricScale.model_validate(_require_dict(item)) for item in data.get("scales", [])],
            disqualifiers=[
                RubricDisqualifier.model_validate(_require_dict(item)) for item in data.get("disqualifiers", [])
            ],
            evidence_requirements=_string_list(data.get("evidence_requirements")),
            output_constraints=_string_list(data.get("output_constraints")),
            decision_thresholds=DecisionThresholds.model_validate(_optional_dict(data.get("decision_thresholds"))),
        )


@dataclass(frozen=True, slots=True)
class RubricFinding:
    code: str
    severity: _FINDING_SEVERITY
    message: str
    path: str = ""


@dataclass(frozen=True, slots=True)
class CompiledRubric:
    schema_version: int
    rubric_id: str
    criterion_ids: list[str]
    result_dimension_ids: list[str]
    scale_ids: list[str]
    normalized_weights: dict[str, float]
    findings: list[RubricFinding]
    prompt_contract: str

    def to_summary(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "rubric_id": self.rubric_id,
            "criterion_ids": list(self.criterion_ids),
            "result_dimension_ids": list(self.result_dimension_ids),
            "scale_ids": list(self.scale_ids),
            "normalized_weights": dict(self.normalized_weights),
            "finding_codes": sorted(finding.code for finding in self.findings),
        }


@dataclass(frozen=True, slots=True)
class RubricPatch:
    op: Literal["append", "replace"]
    path: str
    value: str
    reason: str = ""


@dataclass(frozen=True, slots=True)
class RubricPatchProposal:
    patches: list[RubricPatch]
    requires_human_review: bool
    metrics: dict[str, float]


def legacy_rubric_spec(rubric: str) -> RubricSpec:
    text = rubric.strip()
    return RubricSpec(
        rubric_id="legacy-string-rubric",
        goal=text,
        criteria=[RubricCriterion(id="overall", description=text, scale_id="score", weight=1.0)],
        scales=[RubricScale(id="score", kind="numeric", min_score=0.0, max_score=1.0)],
    )


def lint_rubric_spec(spec: RubricSpec) -> list[RubricFinding]:
    findings: dict[str, RubricFinding] = {}

    def add(code: str, severity: _FINDING_SEVERITY, message: str, path: str = "") -> None:
        findings.setdefault(code, RubricFinding(code=code, severity=severity, message=message, path=path))

    if spec.schema_version != RUBRIC_SPEC_SCHEMA_VERSION:
        add("unsupported_schema_version", "error", "RubricSpec schema_version must be 1", "/schema_version")
    if not spec.criteria:
        add("missing_criteria", "error", "RubricSpec must declare at least one criterion", "/criteria")
    if not spec.scales:
        add("missing_scales", "error", "RubricSpec must declare at least one scale", "/scales")

    criterion_counts = Counter(criterion.id for criterion in spec.criteria)
    for criterion_id, count in criterion_counts.items():
        if count > 1:
            add("duplicate_criterion_id", "error", f"Duplicate criterion id: {criterion_id}", f"/criteria/{criterion_id}")

    scale_counts = Counter(scale.id for scale in spec.scales)
    for scale_id, count in scale_counts.items():
        if count > 1:
            add("duplicate_scale_id", "error", f"Duplicate scale id: {scale_id}", f"/scales/{scale_id}")

    scale_ids = set(scale_counts)
    for index, scale in enumerate(spec.scales):
        if scale.max_score <= scale.min_score:
            add("invalid_scale_range", "error", "Scale max_score must be greater than min_score", f"/scales/{index}")
        if scale.kind == "binary" and (scale.min_score != 0 or scale.max_score != 1):
            add("invalid_binary_scale", "error", "Binary scales must normalize to 0..1", f"/scales/{index}")

    has_scope = bool((spec.scope and spec.scope.has_boundary()) or spec.corpus_profile)
    for index, criterion in enumerate(spec.criteria):
        if criterion.scale_id not in scale_ids:
            add("unknown_scale", "error", f"Criterion {criterion.id} references unknown scale", f"/criteria/{index}/scale_id")
        if criterion.weight <= 0:
            add("invalid_weight", "error", "Criterion weight must be positive", f"/criteria/{index}/weight")
        if criterion.scope and criterion.scope.has_boundary():
            has_scope = True
        generic_words = [word for word in re.split(r"\W+", criterion.description.lower()) if word in _GENERIC_WORDS]
        if len(generic_words) > 2:
            add(
                "vague_criterion",
                "warning",
                "Criterion uses repeated generic terms",
                f"/criteria/{index}/description",
            )

    if not has_scope:
        add("missing_scope_boundaries", "warning", "Rubric has no explicit scope boundaries", "/scope")

    if any(phrase in spec.goal.lower() for phrase in _HYPOTHESIS_LOADED):
        add("hypothesis_loaded_goal", "warning", "Goal appears to load the desired conclusion", "/goal")

    for index, constraint in enumerate(spec.output_constraints):
        if "xml" in constraint.lower():
            add(
                "unsupported_output_constraint",
                "warning",
                "XML-only output is not supported by the judge parser",
                f"/output_constraints/{index}",
            )

    disqualifier_counts = Counter(disqualifier.id for disqualifier in spec.disqualifiers)
    for disqualifier_id, count in disqualifier_counts.items():
        if count > 1:
            add("duplicate_disqualifier_id", "error", f"Duplicate disqualifier id: {disqualifier_id}")

    return sorted(findings.values(), key=lambda finding: finding.code)


def compile_rubric_spec(spec: RubricSpec | str | dict[str, Any]) -> CompiledRubric:
    rubric_spec = _coerce_spec(spec)
    findings = lint_rubric_spec(rubric_spec)
    errors = [finding.code for finding in findings if finding.severity == "error"]
    if errors:
        raise ValueError(f"invalid rubric: {', '.join(sorted(errors))}")

    total_weight = sum(criterion.weight for criterion in rubric_spec.criteria) or 1.0
    normalized_weights = {
        criterion.id: round(criterion.weight / total_weight, 6) for criterion in rubric_spec.criteria
    }
    criterion_ids = [criterion.id for criterion in rubric_spec.criteria]
    scale_ids = list(dict.fromkeys(scale.id for scale in rubric_spec.scales))
    prompt_contract = render_rubric_prompt(rubric_spec)
    return CompiledRubric(
        schema_version=RUBRIC_SPEC_SCHEMA_VERSION,
        rubric_id=rubric_spec.rubric_id,
        criterion_ids=criterion_ids,
        result_dimension_ids=criterion_ids,
        scale_ids=scale_ids,
        normalized_weights=normalized_weights,
        findings=findings,
        prompt_contract=prompt_contract,
    )


def render_rubric_prompt(spec: RubricSpec | CompiledRubric | str | dict[str, Any]) -> str:
    if isinstance(spec, CompiledRubric):
        return spec.prompt_contract
    rubric_spec = _coerce_spec(spec)
    lines = [f"RubricSpec {rubric_spec.rubric_id}", f"Goal: {rubric_spec.goal}", "Criteria:"]
    for criterion in rubric_spec.criteria:
        lines.append(
            f"- {criterion.id} (weight {criterion.weight:g}, scale {criterion.scale_id}): {criterion.description}"
        )
    if rubric_spec.scope and rubric_spec.scope.has_boundary():
        lines.append(f"Scope include: {', '.join(rubric_spec.scope.include) or 'n/a'}")
        lines.append(f"Scope exclude: {', '.join(rubric_spec.scope.exclude) or 'n/a'}")
    if rubric_spec.corpus_profile:
        lines.append(f"Corpus domain: {rubric_spec.corpus_profile.domain}")
        lines.append(f"Corpus source: {rubric_spec.corpus_profile.source_summary}")
    if rubric_spec.disqualifiers:
        lines.append("Disqualifiers:")
        for disqualifier in rubric_spec.disqualifiers:
            lines.append(f"- {disqualifier.id}: {disqualifier.description}")
    if rubric_spec.evidence_requirements:
        lines.append("Evidence requirements: " + "; ".join(rubric_spec.evidence_requirements))
    if rubric_spec.output_constraints:
        lines.append("Output constraints: " + "; ".join(rubric_spec.output_constraints))
    lines.append("Result dimensions: " + ", ".join(criterion.id for criterion in rubric_spec.criteria))
    return "\n".join(lines)


def propose_rubric_patches(
    spec: RubricSpec,
    anchors: list[dict[str, Any]],
    *,
    experimental: bool = False,
) -> RubricPatchProposal:
    if not experimental:
        raise ValueError("rubric patch proposals are experimental")

    grouped: dict[str, list[dict[str, Any]]] = defaultdict(list)
    human_scores: list[float] = []
    judge_scores: list[float] = []
    for anchor in anchors:
        criterion_id = str(anchor.get("criterion_id", ""))
        if criterion_id:
            grouped[criterion_id].append(anchor)
        if anchor.get("human_score") is not None and anchor.get("judge_score") is not None:
            human_scores.append(float(anchor["human_score"]))
            judge_scores.append(float(anchor["judge_score"]))

    patches: list[RubricPatch] = []
    known_ids = {criterion.id for criterion in spec.criteria}
    for criterion_id, items in grouped.items():
        if criterion_id not in known_ids or len(items) < 2:
            continue
        mean_gap = mean(
            abs(float(item.get("judge_score", 0.0)) - float(item.get("human_score", 0.0))) for item in items
        )
        if mean_gap < 0.2:
            continue
        note = _first_note(items)
        patches.append(
            RubricPatch(
                op="append",
                path=f"/criteria/{criterion_id}/description",
                value=f" Calibration note: {note}",
                reason=f"human anchors disagree with judge by {mean_gap:.3f}",
            )
        )

    mean_error = mean(abs(j - h) for h, j in zip(human_scores, judge_scores, strict=True)) if human_scores else 0.0
    metrics = {
        "agreement": round(1.0 - mean_error, 6),
        "consistency": round(1.0 - _score_range(judge_scores), 6),
        "discrimination": round(_score_range(human_scores), 6),
    }
    return RubricPatchProposal(patches=patches, requires_human_review=True, metrics=metrics)


def _coerce_spec(spec: RubricSpec | str | dict[str, Any]) -> RubricSpec:
    if isinstance(spec, RubricSpec):
        return spec
    if isinstance(spec, str):
        return legacy_rubric_spec(spec)
    return RubricSpec.model_validate(spec)


def _string_list(value: Any) -> list[str]:
    if value is None:
        return []
    if isinstance(value, list):
        return [str(item) for item in value]
    return [str(value)]


def _optional_dict(value: Any) -> dict[str, Any] | None:
    return value if isinstance(value, dict) else None


def _require_dict(value: Any) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise TypeError(f"expected object, got {type(value).__name__}")
    return value


def _optional_float(value: Any) -> float | None:
    return None if value is None else float(value)


def _score_range(scores: list[float]) -> float:
    return max(scores) - min(scores) if scores else 0.0


def _first_note(items: list[dict[str, Any]]) -> str:
    for item in items:
        note = str(item.get("human_notes", "")).strip()
        if note:
            return note
    return "tighten this criterion against human anchors"

File: ts/src/judge/rubric-spec.ts
import { z } from "zod";

export const RUBRIC_SPEC_SCHEMA_VERSION = 1;

type FindingSeverity = "warning" | "error";
const genericWords = new Set(["good", "nice", "appropriate", "adequate", "proper"]);
const hypothesisLoaded = ["prove that", "confirm that", "must show", "obviously"];

export const RubricScopeSchema = z
  .object({
    include: z.array(z.string()).default([]),
    exclude: z.array(z.string()).default([]),
  })
  .default({ include: [], exclude: [] });

export const CorpusProfileSchema = z.object({
  domain: z.string().default(""),
  audience: z.string().default(""),
  source_summary: z.string().default(""),
});

export const RubricScaleSchema = z.object({
  id: z.string(),
  kind: z.enum(["numeric", "binary"]),
  min_score: z.number().default(0),
  max_score: z.number().default(1),
  pass_score: z.number().optional(),
  anchors: z.record(z.string()).default({}),
});

export const RubricCriterionSchema = z.object({
  id: z.string(),
  description: z.string(),
  scale_id: z.string(),
  weight: z.number().default(1),
  scope: RubricScopeSchema.optional(),
  evidence_requirements: z.array(z.string()).default([]),
});

export const RubricDisqualifierSchema = z.object({
  id: z.string(),
  description: z.string(),
});

export const DecisionThresholdsSchema = z.object({
  pass_score: z.number().default(0.8),
  excellent_score: z.number().default(0.9),
});

export const RubricSpecSchema = z.object({
  schema_version: z.literal(1).default(1),
  rubric_id: z.string(),
  title: z.string().default(""),
  goal: z.string(),
  metadata: z.record(z.unknown()).default({}),
  scope: RubricScopeSchema.optional(),
  corpus_profile: CorpusProfileSchema.optional(),
  criteria: z.array(RubricCriterionSchema),
  scales: z.array(RubricScaleSchema),
  disqualifiers: z.array(RubricDisqualifierSchema).default([]),
  evidence_requirements: z.array(z.string()).default([]),
  output_constraints: z.array(z.string()).default([]),
  decision_thresholds: DecisionThresholdsSchema.optional(),
});

export type RubricScope = z.infer<typeof RubricScopeSchema>;
export type CorpusProfile = z.infer<typeof CorpusProfileSchema>;
export type RubricScale = z.infer<typeof RubricScaleSchema>;
export type RubricCriterion = z.infer<typeof RubricCriterionSchema>;
export type RubricDisqualifier = z.infer<typeof RubricDisqualifierSchema>;
export type DecisionThresholds = z.infer<typeof DecisionThresholdsSchema>;
export type RubricSpec = z.infer<typeof RubricSpecSchema>;

export type RubricFinding = {
  code: string;
  severity: FindingSeverity;
  message: string;
  path: string;
};

export class CompiledRubric {
  constructor(
    readonly schema_version: 1,
    readonly rubric_id: string,
    readonly criterion_ids: string[],
    readonly result_dimension_ids: string[],
    readonly scale_ids: string[],
    readonly normalized_weights: Record<string, number>,
    readonly findings: RubricFinding[],
    readonly prompt_contract: string,
  ) {}

  toSummary(): Record<string, unknown> {
    return {
      schema_version: this.schema_version,
      rubric_id: this.rubric_id,
      criterion_ids: [...this.criterion_ids],
      result_dimension_ids: [...this.result_dimension_ids],
      scale_ids: [...this.scale_ids],
      normalized_weights: { ...this.normalized_weights },
      finding_codes: this.findings.map((finding) => finding.code).sort(),
    };
  }
}

export type RubricPatch = {
  op: "append" | "replace";
  path: string;
  value: string;
  reason: string;
};

export type RubricPatchProposal = {
  patches: RubricPatch[];
  requires_human_review: boolean;
  metrics: Record<string, number>;
};

export function legacyRubricSpec(rubric: string): RubricSpec {
  const text = rubric.trim();
  return {
    schema_version: 1,
    rubric_id: "legacy-string-rubric",
    title: "",
    goal: text,
    metadata: {},
    criteria: [{ id: "overall", description: text, scale_id: "score", weight: 1, evidence_requirements: [] }],
    scales: [{ id: "score", kind: "numeric", min_score: 0, max_score: 1, anchors: {} }],
    disqualifiers: [],
    evidence_requirements: [],
    output_constraints: [],
  };
}

export function lintRubricSpec(input: RubricSpec): RubricFinding[] {
  const spec = RubricSpecSchema.parse(input);
  const findings = new Map<string, RubricFinding>();
  const add = (code: string, severity: FindingSeverity, message: string, path = ""): void => {
    if (!findings.has(code)) findings.set(code, { code, severity, message, path });
  };

  if (spec.criteria.length === 0) {
    add("missing_criteria", "error", "RubricSpec must declare at least one criterion", "/criteria");
  }
  if (spec.scales.length === 0) {
    add("missing_scales", "error", "RubricSpec must declare at least one scale", "/scales");
  }

  const criterionCounts = countBy(spec.criteria.map((criterion) => criterion.id));
  for (const [criterionId, count] of Object.entries(criterionCounts)) {
    if (count > 1) {
      add("duplicate_criterion_id", "error", `Duplicate criterion id: ${criterionId}`, `/criteria/${criterionId}`);
    }
  }

  const scaleCounts = countBy(spec.scales.map((scale) => scale.id));
  for (const [scaleId, count] of Object.entries(scaleCounts)) {
    if (count > 1) add("duplicate_scale_id", "error", `Duplicate scale id: ${scaleId}`, `/scales/${scaleId}`);
  }

  const scaleIds = new Set(Object.keys(scaleCounts));
  spec.scales.forEach((scale, index) => {
    if (scale.max_score <= scale.min_score) {
      add("invalid_scale_range", "error", "Scale max_score must be greater than min_score", `/scales/${index}`);
    }
    if (scale.kind === "binary" && (scale.min_score !== 0 || scale.max_score !== 1)) {
      add("invalid_binary_scale", "error", "Binary scales must normalize to 0..1", `/scales/${index}`);
    }
  });

  let hasScope = Boolean(hasScopeBoundary(spec.scope) || spec.corpus_profile);
  spec.criteria.forEach((criterion, index) => {
    if (!scaleIds.has(criterion.scale_id)) {
      add("unknown_scale", "error", `Criterion ${criterion.id} references unknown scale`, `/criteria/${index}/scale_id`);
    }
    if (criterion.weight <= 0) add("invalid_weight", "error", "Criterion weight must be positive", `/criteria/${index}/weight`);
    if (hasScopeBoundary(criterion.scope)) hasScope = true;
    const genericCount = criterion.description
      .toLowerCase()
      .split(/\W+/)
      .filter((word) => genericWords.has(word)).length;
    if (genericCount > 2) {
      add("vague_criterion", "warning", "Criterion uses repeated generic terms", `/criteria/${index}/description`);
    }
  });

  if (!hasScope) add("missing_scope_boundaries", "warning", "Rubric has no explicit scope boundaries", "/scope");
  if (hypothesisLoaded.some((phrase) => spec.goal.toLowerCase().includes(phrase))) {
    add("hypothesis_loaded_goal", "warning", "Goal appears to load the desired conclusion", "/goal");
  }
  spec.output_constraints.forEach((constraint, index) => {
    if (constraint.toLowerCase().includes("xml")) {
      add(
        "unsupported_output_constraint",
        "warning",
        "XML-only output is not supported by the judge parser",
        `/output_constraints/${index}`,
      );
    }
  });

  const disqualifierCounts = countBy(spec.disqualifiers.map((disqualifier) => disqualifier.id));
  for (const [disqualifierId, count] of Object.entries(disqualifierCounts)) {
    if (count > 1) add("duplicate_disqualifier_id", "error", `Duplicate disqualifier id: ${disqualifierId}`);
  }

  return [...findings.values()].sort((a, b) => a.code.localeCompare(b.code));
}

export function compileRubricSpec(input: RubricSpec | string | Record<string, unknown>): CompiledRubric {
  const spec = coerceSpec(input);
  const findings = lintRubricSpec(spec);
  const errors = findings.filter((finding) => finding.severity === "error").map((finding) => finding.code).sort();
  if (errors.length) throw new Error(`invalid rubric: ${errors.join(", ")}`);

  const totalWeight = spec.criteria.reduce((total, criterion) => total + criterion.weight, 0) || 1;
  const normalizedWeights = Object.fromEntries(
    spec.criteria.map((criterion) => [criterion.id, round6(criterion.weight / totalWeight)]),
  );
  const criterionIds = spec.criteria.map((criterion) => criterion.id);
  const scaleIds = [...new Set(spec.scales.map((scale) => scale.id))];
  return new CompiledRubric(
    1,
    spec.rubric_id,
    criterionIds,
    criterionIds,
    scaleIds,
    normalizedWeights,
    findings,
    renderRubricPrompt(spec),
  );
}

export function renderRubricPrompt(input: RubricSpec | CompiledRubric | string | Record<string, unknown>): string {
  if (input instanceof CompiledRubric) return input.prompt_contract;
  const spec = coerceSpec(input);
  const lines = [`RubricSpec ${spec.rubric_id}`, `Goal: ${spec.goal}`, "Criteria:"];
  for (const criterion of spec.criteria) {
    lines.push(`- ${criterion.id} (weight ${criterion.weight}, scale ${criterion.scale_id}): ${criterion.description}`);
  }
  if (hasScopeBoundary(spec.scope)) {
    lines.push(`Scope include: ${spec.scope?.include.join(", ") || "n/a"}`);
    lines.push(`Scope exclude: ${spec.scope?.exclude.join(", ") || "n/a"}`);
  }
  if (spec.corpus_profile) {
    lines.push(`Corpus domain: ${spec.corpus_profile.domain}`);
    lines.push(`Corpus source: ${spec.corpus_profile.source_summary}`);
  }
  if (spec.disqualifiers.length) {
    lines.push("Disqualifiers:");
    for (const disqualifier of spec.disqualifiers) lines.push(`- ${disqualifier.id}: ${disqualifier.description}`);
  }
  if (spec.evidence_requirements.length) lines.push(`Evidence requirements: ${spec.evidence_requirements.join("; ")}`);
  if (spec.output_constraints.length) lines.push(`Output constraints: ${spec.output_constraints.join("; ")}`);
  lines.push(`Result dimensions: ${spec.criteria.map((criterion) => criterion.id).join(", ")}`);
  return lines.join("\n");
}

export function proposeRubricPatches(
  specInput: RubricSpec,
  anchors: Array<Record<string, unknown>>,
  opts: { experimental?: boolean } = {},
): RubricPatchProposal {
  if (!opts.experimental) throw new Error("rubric patch proposals are experimental");

  const spec = RubricSpecSchema.parse(specInput);
  const knownIds = new Set(spec.criteria.map((criterion) => criterion.id));
  const grouped = new Map<string, Array<Record<string, unknown>>>();
  const humanScores: number[] = [];
  const judgeScores: number[] = [];

  for (const anchor of anchors) {
    const criterionId = String(anchor.criterion_id ?? "");
    if (criterionId) grouped.set(criterionId, [...(grouped.get(criterionId) ?? []), anchor]);
    if (typeof anchor.human_score === "number" && typeof anchor.judge_score === "number") {
      humanScores.push(anchor.human_score);
      judgeScores.push(anchor.judge_score);
    }
  }

  const patches: RubricPatch[] = [];
  for (const [criterionId, items] of grouped.entries()) {
    if (!knownIds.has(criterionId) || items.length < 2) continue;
    const meanGap = average(items.map((item) => Math.abs(Number(item.judge_score ?? 0) - Number(item.human_score ?? 0))));
    if (meanGap < 0.2) continue;
    patches.push({
      op: "append",
      path: `/criteria/${criterionId}/description`,
      value: ` Calibration note: ${firstNote(items)}`,
      reason: `human anchors disagree with judge by ${meanGap.toFixed(3)}`,
    });
  }

  const meanError = average(humanScores.map((score, index) => Math.abs((judgeScores[index] ?? 0) - score)));
  return {
    patches,
    requires_human_review: true,
    metrics: {
      agreement: round6(1 - meanError),
      consistency: round6(1 - scoreRange(judgeScores)),
      discrimination: round6(scoreRange(humanScores)),
    },
  };
}

function coerceSpec(input: RubricSpec | string | Record<string, unknown>): RubricSpec {
  if (typeof input === "string") return legacyRubricSpec(input);
  return RubricSpecSchema.parse(input);
}

function countBy(values: string[]): Record<string, number> {
  const counts: Record<string, number> = {};
  for (const value of values) counts[value] = (counts[value] ?? 0) + 1;
  return counts;
}

function hasScopeBoundary(scope?: RubricScope): boolean {
  return Boolean(scope && (scope.include.length > 0 || scope.exclude.length > 0));
}

function round6(value: number): number {
  return Math.round(value * 1_000_000) / 1_000_000;
}

function average(values: number[]): number {
  return values.length ? values.reduce((total, value) => total + value, 0) / values.length : 0;
}

function scoreRange(values: number[]): number {
  return values.length ? Math.max(...values) - Math.min(...values) : 0;
}

function firstNote(items: Array<Record<string, unknown>>): string {
  for (const item of items) {
    const note = String(item.human_notes ?? "").trim();
    if (note) return note;
  }
  return "tighten this criterion against human anchors";
}
